Dense->BSR performance improvment

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82121
* #82122
* #83086
* __->__ #83085
* #83084

Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

amjames committed:
Dense->BSR performance improvment

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

bhosmer reviewed (approved):
Not just a perf improvement but (going by the deleted comments) also relaxes the same-sparsity-pattern restriction, right? Which if so is as big a deal as decent perf I'd say

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

nikitaved reviewed (approved):
Awesome! Faster and less code - double win!

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]

amjames commented:
@pytorchbot merge

amjames committed:
Update on "Dense->BSR performance improvment"


Applies the algorithm for re-batching compressed indices to avoid n-batch kernel launches. This is an optimization for `dim()>= 3` inputs and does not change behavior in any way.

[ghstack-poisoned]
